src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -24,133 +24,124 @@ trait Batchable {
- * this optimization since they don't know whether the underlying
- * executor will benefit from it.
- * A batching executor can create deadlocks if code does
- * not use `scala.concurrent.blocking` when it should,
- * because tasks created within other tasks will block
- * on the outer task completing.
- * This executor may run tasks in any order, including LIFO order.
- * There are no ordering guarantees.
- *
- * WARNING: The underlying Executor's execute-method must not execute the submitted Runnable
- * in the calling thread synchronously. It must enqueue/handoff the Runnable.
- */
- private[concurrent] trait BatchingExecutor extends Executor {
-  private[this] final val _tasksLocal = new ThreadLocal[Batch]()
+private[concurrent] trait BatchingExecutor extends Executor {
+  private[this] final val _tasksLocal = new ThreadLocal[AnyRef]()
 
-  private[this] final class Batch extends Runnable with BlockContext with (BlockContext => Throwable) {
+  /*
+   * Batch implements a LIFO queue (stack) and is used as a trampolining Runnable.
+   * In order to conserve allocations, the first element in the batch is stored "unboxed" in
+   * the `first` field. Subsequent Runnables are stored in the array called `other`.
+  */
+  private[this] final class Batch(private[this] final val resubmitOnBlock: Boolean) extends Runnable with BlockContext with (BlockContext => Throwable) {
     private[this] final var parentBlockContext: BlockContext = _
     private[this] final var first: Runnable = _
     private[this] final var size: Int = _
     private[this] final var other: Array[Runnable] = BatchingExecutorStatics.emptyBatchArray
 
-    def this(r: Runnable) = {
-      this()
+    def this(r: Runnable, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       first = r
       size = 1
     }
 
-    private def this(first: Runnable, other: Array[Runnable], size: Int) = {
-      this()
+    private def this(first: Runnable, other: Array[Runnable], size: Int, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       this.first = first
       this.other = other
       this.size = size
     }
 
     private[this] final def cloneAndClear(): Batch = {
-      val newBatch = new Batch(first, other, size)
+      val newBatch = new Batch(first, other, size, resubmitOnBlock)
       this.first = null
       this.other = BatchingExecutorStatics.emptyBatchArray
       this.size = 0
       newBatch
     }
 
-    private[this] final def grow(): Unit = {
-      val len = other.length
-      other =
-        if (len == 0) new Array[Runnable](4)
-        else {
-          val newOther = new Array[Runnable](len << 1)
-          System.arraycopy(other, 0, newOther, 0, len)
-          newOther
-        }
+    private[this] final def ensureCapacity(curSize: Int): Array[Runnable] = {
+      val curOther = this.other
+      val curLen = curOther.length
+      if (curSize <= curLen) curOther
+      else {
+        val newLen = if (curLen == 0) 4 else curLen << 1
```

Comment:
I investigated different growth strategies, but here's the thing—if tasks are added incrementally they tend to be added faster than they can be consumed, which means that we need to grow fast anyway.